crates/pathfinder/src/sync.rs

```diff
@@ -0,0 +1,300 @@
+#![allow(dead_code, unused_variables)]
+
+use std::time::Duration;
+
+use anyhow::{Context, Result};
+use futures::{Future, FutureExt};
+use pathfinder_common::{felt, felt_bytes, BlockHash, BlockNumber};
+use tokio::task::JoinHandle;
+
+pub struct BlockHeader {
+    // TODO
+}
+
+pub struct BlockBody {
+    // TODO
+}
+
+pub struct StateUpdate {
+    // TODO
+}
+
+#[allow(unused_variables)]
+#[async_trait::async_trait]
+pub trait Source {
+    async fn block_headers(&self, range: std::ops::Range<BlockNumber>) -> Result<Vec<BlockHeader>> {
+        unimplemented!();
+    }
+
+    async fn block_bodies(&self, range: std::ops::Range<BlockNumber>) -> Result<Vec<BlockBody>> {
+        unimplemented!();
+    }
+
+    async fn state_updates(&self, range: std::ops::Range<BlockNumber>) -> Result<Vec<StateUpdate>> {
+        unimplemented!();
+    }
+}
+
+pub type NewHeaderSource = tokio::sync::mpsc::Receiver<BlockHeader>;
+
+enum Mode {
+    Tracking,
+    Syncing,
+}
+
+async fn run(source: impl Source) -> Result<()> {
+    // L1 new head
```

Comment:
Wow, there are a lot of comments. Maybe extract it into the design doc? (Just a suggestion).